### ISSUE TITLE
BooleanCircuit: Templates declared in header, defined in implementation

### ISSUE THREAD

**Issue by lenerd** [encryptogroup/ABY#77]
Various methods of `Booleancircuit` (e.g. `PutINGate`) are declared as templates in `booleancircuits.h`. The corresponding template definition is located in `booleancircuits.cpp` and, thus, not visible for uses of the class. This results in linker errors if a method is used but not instantiated for a particular template parameter.

**Comment by lenerd** [comment]
~The following template functions are declared but not implemented.~
```c++
template<class T> share* InternalPutINGate(uint32_t nvals, T val, uint32_t bitlen, e_role role);
template<class T> share* InternalPutSharedINGate(uint32_t nvals, T val, uint32_t bitlen);
```
~So I guess the functions ```PutINGate(uint64_t val, uint32_t bitlen, e_role role)``` etc. using these templates probably have never worked.~

Edit: My mistake, I found them ...

**Comment by lenerd** [comment]
fixed by #87
